src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,251 @@
+    {
+        #region Constants
+        private const string UriSchemeHttps = "https";
+        #endregion
+
+        #region Fields
+        private volatile bool _anyOperationStarted;
+        private volatile bool _disposed;
+        private bool _automaticRedirection = true;
+        #endregion
+
+        internal CurlHandler()
+        {
+        }
+
+        #region Properties
+        internal bool AutomaticRedirection
+        {
+            get
+            {
+                return _automaticRedirection;
+            }
+
+            set
+            {
+                CheckDisposedOrStarted();
+                _automaticRedirection = value;
+            }
+        }
+        #endregion
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+            }
+
+            base.Dispose(disposing);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+            if (request.RequestUri.Scheme == UriSchemeHttps)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTPS stack is not yet implemented");
+            }
+            if (request.Content != null)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTP requests with a body are not yet supported");
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+
+            try
+            {
+                Task.Factory.StartNew(
+                    StartRequest,
+                    state,
+                    CancellationToken.None,
+                    TaskCreationOptions.DenyChildAttach,
+                    TaskScheduler.Default);
+            }
+            catch (Exception ex)
+            {
+                throw new HttpRequestException("SendAsync", ex);
+            }
+
+            return state.Task;
+        }
+
+        #region Private methods
+
+        private void StartRequest(object obj)
+        {
+            RequestState state = (RequestState)obj;
+            HttpResponseMessage responseMessage = null;
+            Exception savedException = null;
+            SafeCurlHandle requestHandle = null;
+
+            if (state.CancellationToken.IsCancellationRequested)
+            {
+                state.TrySetCanceled();
+                return;
+            }
+
+            state.CancellationToken.Register(x => ((RequestState)x).TrySetCanceled(), state);
```

Comment:
No, don't revert it.  I'm saying that you need to store the result of this into a local or a field, e.g.
```C#
CancellationTokenRegistration registration = state.CancellationToken.Register(...);
```
and then later on when this request is done and you no longer want cancellation registered, dispose of it:
```C#
registration.Dispose();
```
You need to do that regardless of the overload of CancellationToken.Register you're using.